nni/algorithms/compression/v2/pytorch/utils/evaluator.py

```diff
@@ -0,0 +1,581 @@
+
+
+class ModuleHook(Hook):
+    def __init__(self, target: Module, target_name: str, hook_factory: Callable[[List], Callable[[Module, Tensor, Tensor], Any]]):
+        assert isinstance(target, Module)
+        super().__init__(target, target_name, hook_factory)
+
+
+class ForwardHook(ModuleHook):
+    def _register(self, hook_func: Callable[[Module, Tensor, Tensor], Any]):
+        return self.target.register_forward_hook(hook_func)  # type: ignore
+
+
+class BackwardHook(ModuleHook):
+    def _register(self, hook_func: Callable[[Module, Tensor, Tensor], Any]):
+        return self.target.register_backward_hook(hook_func)  # type: ignore
+
+
+class Evaluator:
+    def _init_optimizer_helpers(self, pure_model: Module | pl.LightningModule):
+        # NOTE: this is a part of Evaluator initialization, please make sure this function has been called before using evaluator.
+        raise NotImplementedError
+
+    def bind_model(self, model: Module | pl.LightningModule, param_names_map: Dict[str, str] | None = None):
+        # param_names_map maps the names of the parameters in the pure_model to the names of the parameters in the binded model.
+        # The format of param_names_map is {pure_model_param_name: binded_model_param_name}.
+        # param_names_map is for initializing the optimizers for the binded model.
+        raise NotImplementedError
+
+    def unbind_model(self):
+        # Evaluator can be reused by `unbind_model` then bind a new model by `bind_model`.
+        raise NotImplementedError
+
+    def patch_loss(self, patch: Callable[[Tensor], Tensor]):
+        raise NotImplementedError
+
+    def revert_loss(self):
+        raise NotImplementedError
+
+    def patch_optimizer_step(self, before_step_tasks: List[Callable], after_step_tasks: List[Callable]):
+        # Run tasks in `before_step_tasks` before `optimizer.step()` each time.
+        # Run tasks in `after_step_tasks` after `optimizer.step()` each time.
+        # NOTE: we only patch these tasks to the first optimizer right now.
+        raise NotImplementedError
+
+    def revert_optimizer_step(self):
+        raise NotImplementedError
+
+    def register_hooks(self, hooks: List[Hook]):
+        raise NotImplementedError
+
+    def remove_all_hooks(self):
+        raise NotImplementedError
+
+    def train(self, max_steps: int | None = None, max_epochs: int | None = None):
+        raise NotImplementedError
+
+    def finetune(self):
+        raise NotImplementedError
+
+    def evaluate(self) -> float | Tuple[float, Any]:
+        # Note that the first item of the returned value will be used as the default metric used by NNI.
+        raise NotImplementedError
+
+    def get_dummy_input(self) -> Any:
+        raise NotImplementedError
+
+
+class LightningEvaluator(Evaluator):
+    """
+    Parameters
+    ----------
+    trainer
+        Pytorch-Lightning Trainer.
+        If the test metric is needed by nni, please make sure log metric with key ``NNI_METRIC``.
```

Comment:
Looks like a sentence appearing in a wrong place.

Reply:
move to class docstring